### ISSUE TITLE
option to disable sleepcover magnet actions?

### ISSUE THREAD

**Issue by WrinklyWink** [koreader/koreader#1382]
If I carry the Kobo Aura with my microsoft universal mobile keyboard the magnets in the latches causes the kobo to awaken and waste power. Is there an option to disable the sleepcover functionality? Can one be implemented?

**Comment by hwhw** [comment]
I don't think that would be easy to do. Or maybe pretty easy, but not using Koreader...
What we actually can and should do is to implement a timer to go back into standby mode after a certain amount of time without any events.

**Comment by Markismus** [comment]
The option to disable is in Kobo's Nickel>options>energy>sleepcover>on/off.

**Comment by WrinklyWink** [comment]
hwhw, you're right. I'm surprised a timer hasn't been implemented over the other extraneous features..
Markismus, this option does not stand globally, that is, it doesn't hold true in Koreader.

**Comment by Frenzie** [comment]
@hwhw Are you sure? It seems to be handled separately from regular power events.
https://github.com/koreader/koreader/blob/a5cff2c42b580946e0e75b2cbfc86c4ea0a7cf2f/frontend/device/input.lua#L239-L246

**Comment by Markismus** [comment]
That's new....as in it wasn't like that a year ago. 

What I fail to understand is: why doesn't koreader return to standby after being woken up. The magnet is still in place: magnet_in_place==standby.

**Comment by WrinklyWink** [comment]
Frenzie, It isn't, I just tested it now after confirming it's option disabled in nickel. I even tried deleting that entry among others to break the support but then it causes standby and page movements incompatibilities. My guess is its because koreader works in a fashion of working over some other primary system, like using a secondary css to completely change how a page is structured.
Markismus, actually because the magnet isn't really on top of the screen and more "near" it, it has a habit of pulsing koreader out of standby. Luckily the magnet sensor is only on the right side of the aura so if I place that side opposite to the magnetic latch there aren't problems. But still this is a primitive work around.

**Comment by Markismus** [comment]
Still saying: magnet should equal standby. How are those events structured? If all values are generated: 
```lua
-- key press event values (KEY.value)
local EVENT_VALUE_KEY_PRESS = 1
local EVENT_VALUE_KEY_REPEAT = 2
local EVENT_VALUE_KEY_RELEASE = 0
```
, couldn't we use more values to get a better response? @WrinklyWink Could you generate a log both for clean sleepcover events and for "pulsing" events? (If you run koreader in debug mode the file `ev.log` is generated. Just add the `-d` flag afer `reader.lua` in the koreader sh-script.)

**Comment by Markismus** [comment]
@WrinklyWink How are you starting koreader? If you are using KSM no other primary systems should be enabled anymore, right?

**Comment by hwhw** [comment]
@Frenzie that, however, doesn't wake it up, I think - it only presents an event? However, I can't currently trace where that event originates (if it actually does fire at all).

@Markismus isn't that Edge-triggered? As I understand, the Kobo isn't stored in a case with a magnet, but the magnet from the Keyboard comes and goes (the latter resuming the Kobo from Standby).

Hm, if the option from Nickel doesn't apply for Koreader, I'm a bit unsure how they do it? Returning from standby isn't triggered by software (as software isn't running then) but rather by hardware, and the place where control resumes to is - in a first step - the kernel. So we probably can flag something that resembles Nickel's behaviour. It'll be a bit harder to find out what that actually is.

**Comment by Frenzie** [comment]
@hwhw Sure, but for all I know the hardware wakes up and Nickel simply goes right back to sleep again. You're just assuming it does it in the way that makes the most sense. ;)

**Comment by Markismus** [comment]
If it's [edge triggered](http://electronics.stackexchange.com/questions/21886/what-does-edge-triggered-and-level-triggered-mean), I still don't know which values are generated when it's active.

**Comment by WrinklyWink** [comment]
Interesting, there are more plausible options than I thought.
@Markismus, I'm starting it up via KSM on the inclination that skipping nickel saves resources and thus battery life. Also my png for loading koreader from the library doesn't go to koreader but to KSM instead. I'm looking up how to debug right now (I've never done it on this device before)

**Comment by hwhw** [comment]
@Frenzie Oh, OK. Well, I consider that a hack, but then if it works....

If it is a "key" that triggers events, we need to look closely if it is reliable. @Markismus is probably right in that our current implementation doesn't really make sense. Or rather, the event naming doesn't make sense. It isn't about standby and resume (those are merely consequences - for standby - or coincidences - for resume), but for the presence (or lack thereof) of magnetism.

**Comment by Markismus** [comment]
If there are two values (on and off), I would assume that any passing magnet has zero effect. If one of the events would be ignored due to resuming/waking up, then we only would have to build in a wait state to check whether the sequence of key events is actually the removing of the battery.

**Comment by hwhw** [comment]
@Markismus except "zero effect" isn't what we want. We want to go back to sleep, as it seems, as we probably don't control whether the device will wake up (if it is correct what is written above and the setting in Nickel merely leads to going back to standby). Now how do we establish that we did wake up (unintended) via passing by of a magnet? The question is whether input events are generated here (as for types of input events: that number is defined via the kernel API, and the "key" number is defined here: https://github.com/koreader/koreader/blob/a5cff2c42b580946e0e75b2cbfc86c4ea0a7cf2f/frontend/device/kobo/device.lua#L66).

**Comment by WrinklyWink** [comment]
@hwhw you can't, because people just want to read their books the very moment the sleep cover is removed which demands instantaneous action. But I was talking about a buffer in the miliseconds..

**Comment by hwhw** [comment]
If we can determine that we wake up due to the passing of a magnet, we can - as Nickel does - implement a configuration setting that will - if set - trigger a return to standby in such a case.

Alternatively or additionally, we can try to be fancy and set that flag (for immediate return to standby) when we went to standby (which is a code path we actually control, as opposed to resuming from standby) by a magnet event.

What buffer were you talking about? I think I'm losing track a bit and this isn't a coherent discussion. The problem at hand, however, seems simple enough. We just need to see what events there are when returning from standby via a magnet event.

**Comment by Frenzie** [comment]
@hwhw I didn't say that's what Nickel does; merely that I don't believe we have any idea just *what* Nickel does unless someone tests it. (NB I don't have a sleep cover.) In any case, @Markismus did point out there are two distinct issues here:

1. Sleepcover disabled.
2. Magnet presence in case of incidental wake-up when enabled.

Perhaps it would be a hack for #1, but it seems to me you could swat two flies at once by simply saying if sleepcover-disabled or magnet-present then suspend.

**Comment by hwhw** [comment]
OK. We need proper debugging of events, otherwise we're simply in the dark here.

**Comment by WrinklyWink** [comment]
whats the syntax for "spoiler" on github? I want to copy and paste it but I don't want to ruin the flow of conversation by having a 2 page scroll of events.

**Comment by Frenzie** [comment]
@WrinklyWink I don't think such a thing exists. I suppose you could always share through a link to pastebin or some such.

**Comment by WrinklyWink** [comment]
[ev.log](http://pastebin.com/YVb0yRt8) of stationary place on top of keyboard (kobo face up, but magnetic latch directly under), and then of keyboard latch pulsing/pass over right side of kobo face

**Comment by Markismus** [comment]
In the 127 lines only the first 4 lines have ev_key 59. The rest is touchscreen input.
```
ev.type	ev.code	ev.value	ev.time.sec	ev.time.usec		ev.type	ev.code	ev.value
1	59	1	1420541339	824547		EV_KEY	59	1
1	59	0	1420541345	126501		EV_KEY	59	0
1	59	1	1420541349	349038		EV_KEY	59	1
1	59	0	1420541349	643433		EV_KEY	59	0
```

So every time a nice pair of events is passed to koreader. If none are ignored, the last one is magnet off, thus getting out of standby.

**Comment by Markismus** [comment]
I actually use this behavior to wake up my mother's Glo, which power button is dead.

**Comment by WrinklyWink** [comment]
does anyone understand how the nickel software toggles this option?

**Comment by Markismus** [comment]
@WrinklyWink Why not try Frenzie's suggestion:
```lua
 -- Kobo sleep
local sleepcover_enabled = false
if keycode == "Power_SleepCover" then
     if ev.value == EVENT_VALUE_KEY_PRESS then
          return "Suspend"
     else
          if sleepcover_enabled then 
               return "Resume" 
          end
     end
end
```

**Comment by hwhw** [comment]
As I wrote above, that does not trigger a resume from standby, but rather generates a resume event in Koreader upon Resume - well, that is not true, actually, it is rather on certain kinds of resumes. All it will do is not to generate the event. The device will wake up nevertheless.

In fact - but this is quite simplified - it should generate a "Suspend" event (yes, "Suspend") when the sleepcover_enabled variable is not set, to trigger a return into Suspend state.

**Comment by hwhw** [comment]
@WrinklyWink - what did you do to generate the events? There's a 6 second interval between the first and a <1 second interval between the second set of events.

**Comment by WrinklyWink** [comment]
@Markismus, it works but it's primitive in the sense where it puts it on standby just by being near the magnetic latch. Complete magnetic disabling is preferred.
@hwhw, I'm having some trouble understanding what you mean exactly. It's because I'm not grasping the logic of the words. Are you implying that the device will still be on and thus draining the battery?
The first event was simple device placement on top of the keyboard while being on. kobo aura was face up while the bluetooth keyboard was underneath with the magnetic latch to the right. I left it there for a bit and then picked it back up. the second was a "pulsing" meaning I ran the latch over the screen very briefly.

**Comment by Markismus** [comment]
Right! So does this work better?

```lua
 -- Kobo sleep
local sleepcover_disabled = true
if keycode == "Power_SleepCover" then
     if ev.value == EVENT_VALUE_KEY_PRESS or sleepcover_disabled then
          return "Suspend"
     else
          return "Resume" 
     end
end
```

or 


```lua
 -- Kobo sleep
local sleepcover_disabled = true
if keycode == "Power_SleepCover" then
     if sleepcover_disabled then return
     elsif ev.value == EVENT_VALUE_KEY_PRESS
          return "Suspend"
     else
          return "Resume" 
     end
end
```

**Comment by hwhw** [comment]
Mean by what? I have a hard time to follow the discussion myself. What I'm almost sure about is this:
* The key events for the "Power_SleepCover" "key" come from the kernel.
* The functionality is in fact implemented like a key. Magnet near sensor = key press, magnet away from sensor = key release. Think of a reed contact instead of a microswitch.

When a magnet comes near the sensor, this is what happens:
* The sensor senses the magnet and communicates with the driver (in-kernel)
* The kernel generates an input event via the input API
* Koreader gets the input event ("Power_SleepCover" pressed)
* Koreader handles that input event by issuing a (Koreader-internal) "Suspend" event here: https://github.com/koreader/koreader/blob/a5cff2c42b580946e0e75b2cbfc86c4ea0a7cf2f/frontend/device/input.lua#L239-L246
* That suspend event is handled within Koreader leads to various things like turning off the frontlight, showing a screensaver and finally triggering the actual suspend by using the Kernel API for this.
* Kernel gains (before the suspend: final) control and puts device into suspend mode.

When the device is in suspend mode and a magnet leaves (thus: edge-triggered) the sensor's area:
* The sensor generates a hardware interrupt, waking the SoC/CPU.
* The kernel gains (back) control and is notified it has just been woken up
* The kernel at this point or shortly after generates an input API event (key released) for the virtual key of the sensor
* Koreader gets control. It's in its event-reading loop and will see the input API event
* Koreader will trigger a Koreader-internal "resume" event (which will remove the screensaver and show whatever else is currently active).

By just killing that last step, you gain not much, as you can see. Koreader will still be running. Probably it will react to some events even. It's not exactly a defined state for now.

I see room for further testing (a) and for optimization (b).

As for (a):
What events are seen when the device is put into suspend by using the power button, then woken up later via passing of a magnet (maybe a real slow pass and a faster pass, so we can see what events are generated)?

As for (b):
* we can't prevent waking up. Really. Hardware is waking up the device, not software (well, this is simplified, but sufficient as an explanation for now). What we can do is to go to sleep again.
* we probably want configurability:
** a flag whether the magnet should be honored at all. If not so, go into suspend immediately again when woken up via magnet (indicated by presence of the input event).
** another flag that will allow "symmetrical" magnet-based suspend: When put into suspend via magnet (cover closing), then wake up when the magnet is gone. When put into sleep via button, do not wake up (i.e. go into suspend again immediately) when magnet is gone/passes.
* we probably don't want instantaneous suspend even when a magnet comes near the sensor, but rather wait, say, 1 second, and only go into suspend when there was no "release" event from Kernel input API during that time.

However, as @Frenzie wrote above, it still is speculation if that is what Nickel does for preventing resume via Magnet. I guess we need a kernel log for Nickel standby events perusing a Magnet, both with the setting enabled and disabled, to be sure. I'm out - I don't have a Kobo with magnet sensor (or does the Mini have one? will check).

**Comment by WrinklyWink** [comment]
@Markismus,
```lua
 -- Kobo sleep
local sleepcover_disabled = true
if keycode == "Power_SleepCover" then
     if ev.value == EVENT_VALUE_KEY_PRESS or sleepcover_disabled then
          return "Suspend"
     else
          return "Resume" 
     end
end
```
yields the same result (meaning still waking from magnet) as:
```lua
 -- Kobo sleep
local sleepcover_enabled = false
if keycode == "Power_SleepCover" then
     if ev.value == EVENT_VALUE_KEY_PRESS then
          return "Suspend"
     else
          if sleepcover_enabled then 
               return "Resume" 
          end
     end
end
```
and
```lua
 -- Kobo sleep
local sleepcover_disabled = true
if keycode == "Power_SleepCover" then
     if sleepcover_disabled then return
     elseif ev.value == EVENT_VALUE_KEY_PRESS
          return "Suspend"
     else
          return "Resume" 
     end
end
```
prevents koreader from booting up, even after rebooting KSM. I tested this twice.

@Markismus, @hwhw, y'all are wizards.

@hwhw, so to be clear, when using the above methods (directly above, 1st and 2nd codes specifically), battery drain quicker if it happens to encounter a magnet? If this is true what makes the above code's convenience arguable is weighing battery-life versus accidental button input :disappointed:, which upon a second thought isn't very likely considering it is a capacitive screen :neutral_face:. But all things aside, if I value battery-life (I do) would the above code have an adverse effect on it? Wouldn't it be better to *know* when it's on in order to then turn it off?

>* we can't prevent waking up. Really. Hardware is waking up the device, not software (well, this is simplified, but sufficient as an explanation for now). What we can do is to go to sleep again.
we probably want configurability:
    * a flag whether the magnet should be honored at all. If not so, go into suspend immediately again when woken up via magnet (indicated by presence of the input event).
    * another flag that will allow "symmetrical" magnet-based suspend: When put into suspend via magnet (cover closing), then wake up when the magnet is gone. When put into sleep via button, do not wake up (i.e. go into suspend again immediately) when magnet is gone/passes.
* we probably don't want instantaneous suspend even when a magnet comes near the sensor, but rather wait, say, 1 second, and only go into suspend when there was no "release" event from Kernel input API during that time.


So this cannot be implemented until we have a kernel log?

>I guess we need a kernel log for Nickel standby events perusing a Magnet, both with the setting enabled and disabled, to be sure.


If you run me through the steps I'll gladly do it :hand:.

>(or does the Mini have one? will check).


Seems unlikely considering they don't advertise that ability on the Kobo website.

**Comment by houqp** [comment]
This feature is already implemented by @Hzj-jie : https://github.com/koreader/koreader/blob/master/frontend/device/kobo/device.lua#L107

We should add the option setting in the up coming koreader home.

**Comment by protist** [comment]
Thanks @houqp. How do we enable this option?

**Comment by houqp** [comment]
@protist just set `ignore_power_sleepcover` to true in your reader setting. You will have to manually edit `settings.reader.lua` file.

**Comment by protist** [comment]
Thanks @houqp. However, I couldn't find `settings.reader.lua` anywhere. I also grepped for `ignore_power_sleepcover`, but I could only find it in the code itself (i.e. `koreader/frontend/device/kobo/device.lua`).

**Comment by protist** [comment]
Having said that, I tested by deleting these [three lines](https://github.com/koreader/koreader/blob/master/frontend/device/kobo/device.lua#L107-109), and it works. The magnet on my cover now does nothing. Obviously it'd be neater if I could find the setting, but the code is definitely working at least.

**Comment by houqp** [comment]
Which version are you using? You should be able to find `settings.reader.lua` under `.kobo/koreader` directory.

**Comment by protist** [comment]
I was using the nightly [koreader-kobo-arm-linux-gnueabihf-v2015.11-483-geb2d4e9.zip](https://github.com/koreader/koreader/releases/tag/v2016.06.24-nightly) from two builds ago. I double-checked but that file is definitely not there. I just downloaded the very latest nightly (+1 build), but it's not there either.

**Comment by houqp** [comment]
The setting file is not included with the release. Instead, it will be created dynamically after you run and exit the reader for the first time. Have you tried exiting the reader before checking the file?

**Comment by protist** [comment]
Oh, I see now, @houqp. Thanks for the tip; I see it now. So eventually this option will end up in `defaults.lua`? Then when users update koreader they can just copy their preferences across in the initial install, rather than having to reboot, then reconnect? As it is, it's simpler for me to just remove those lines. Cheers.

**Comment by houqp** [comment]
@protist how are you updating koreader right now? Looks like you are wiping everything with a clean install everytime? `settings.reader.lua` is supposed to be kept during updates so all your settings will be not wiped during update.

We are moving away from defaults.lua as `settings.reader.lua` is the preferred way of managing user configs.

**Comment by protist** [comment]
@houqp Yes, you are correct. I was wiping everything with a clean install, and just copying `defaults.lua`. The README doesn't really mention upgrading, so I presumed this was the preferred way to do it, but apparently I guessed wrong!

What is the current recommended way of upgrading, then? Just manually copying over both `defaults.lua` and `settings.reader.lua`? (And eventually just the latter?)

**Comment by KenMaltby** [comment]
For Windows users it is simple to "Drag&Drop" the "koreader" folder from WinRAR or 7-zip into the appropriate directory of your device.  You are then offered the option to "Merge" the folders.

**Comment by protist** [comment]
Ah okay. I use Linux, but easy enough to `rsync`. I thought there might be files deleted in newer builds, but I guess they'd just be not referenced (and just take up a minimal amount of space). Thanks for the advice.

**Comment by houqp** [comment]
I basically do the same thing as what @KenMaltby does ;p But I am implementing a proper network manager so I can do the update without having to plug in/out the device all the time, which can be very annoying.

`rsync` is cleaner if you automated it. Also I believe if you update it via OTA update feature, it should delete out of date files for you too.

**Comment by houqp** [comment]
BTW, `defaults.lua` will be overridden by an upgrade. If you have any custom settings in `defaults.lua`, you should put it into `defaults.persistent.lua` instead. If the README is not clear, please feel free to send us pull request to improve it. Or you can just open an issue to let us know which part is confusing so we can fix it.

**Comment by protist** [comment]
@houqp I did not even know if there was an OTA feature… there's nothing in the README, and just a passing reference in the wiki?

Regarding `defaults.persistent.lua`, I think the [wiki](https://github.com/koreader/koreader/wiki/Change-defaults) is clear enough, but again, this refers to the OTA. Also, in my last update, `defaults.lua` had changed, so my previous config file had to be modified manually anyway.

**Comment by houqp** [comment]
There are many hidden features in KOReader that I don't even know ;P

**Comment by KenMaltby** [comment]
The OTA servers provide us users with access to the current "Nightly" (or twice daily?) auto generated "release".  I normally download the current OTA nightly, to my PC and make changes to it then recompress the file and leave it ready to be installed, as I described above.  I tried the OTA process on my AuraHD and it worked fine for 512, yesterday.  I had to connect with the USB afterwards to make my changes.  This time to KOReader files installed on the device.  All seems to be working as expected.
